### PR TITLE
Implement HTMLTableElement#rows

### LINKED ISSUE #8084: Implement <table> 'rows' attribute
https://html.spec.whatwg.org/multipage/tables.html#dom-table-rows

This requires `HTMLCollection` to be updated to support more flexible custom iterators than just a simple filter

I started refactoring `HTMLCollection` and implementing `<table>` `rows` in [this branch](https://github.com/servo/servo/compare/master...frewsxcv:traversal-htmlcollection?expand=1), but I'm going to hold off on contributing more to it for the time being incase anyone thinks of a better system. If anyone wants to give this a shot, feel free to steal my existing work

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #8084.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10086)

### CODE REVIEW

**Comment by KiChjang** [comment]
```
Ran 4618 tests finished in 963.0 seconds.
  • 4614 ran as expected. 1361 tests skipped.
  • 1 tests crashed unexpectedly
  • 3 tests had unexpected subtest results

Tests with unexpected results:
  ▶ CRASH [expected OK] /html/dom/interfaces.html
  │ 
  │ callback not yet supported
  │ callback not yet supported
  │ callback not yet supported
  │ callback not yet supported
  │ callback not yet supported
  │ callback not yet supported
  └ callback not yet supported

  ▶ Unexpected subtest result in /html/semantics/tabular-data/the-caption-element/caption_001.html:
  └ PASS [expected FAIL] caption of the third table element should be null

  ▶ Unexpected subtest result in /html/semantics/tabular-data/the-table-element/insertRow-method-02.html:
  └ PASS [expected FAIL] table should start out empty

  ▶ Unexpected subtest result in /html/semantics/tabular-data/the-table-element/table-rows.html:
  └ PASS [expected FAIL] Children of table

  ▶ Unexpected subtest result in /html/semantics/tabular-data/the-table-element/table-rows.html:
  └ PASS [expected FAIL] Children of thead

  ▶ Unexpected subtest result in /html/semantics/tabular-data/the-table-element/table-rows.html:
  └ PASS [expected FAIL] Children of tbody

  ▶ Unexpected subtest result in /html/semantics/tabular-data/the-table-element/table-rows.html:
  └ PASS [expected FAIL] Children of tfoot
```
I'm not sure what the first crash is about.

File: components/script/dom/htmltableelement.rs [hunk 1]
```diff
@@ -51,6 +54,31 @@ impl HTMLTableElement {
 }
 
 impl HTMLTableElementMethods for HTMLTableElement {
+    // https://html.spec.whatwg.org/multipage/#dom-table-rows
+    fn Rows(&self) -> Root<HTMLCollection> {
+        #[derive(JSTraceable, HeapSizeOf)]
+        #[allow(moved_no_move)]
+        struct TableRowFilter {
```
Comment (Manishearth) [inline components/script/dom/htmltableelement.rs]: Why not tag this with `#[no_move]` instead?

File: components/script/dom/htmltableelement.rs [hunk 2]
```diff
@@ -51,6 +54,31 @@ impl HTMLTableElement {
...
+            sections: RootedVec<JS<Node>>
+        }
+
+        impl CollectionFilter for TableRowFilter {
```
Comment (Manishearth) [inline components/script/dom/htmltableelement.rs]: This is doing more traversal than it needs to, but there doesn't seem to be any way of avoiding it with the HTMLCollection API.

I wonder if we could tweak HTMLCollection to take a general iterator with utility methods to create a filter iterator, cc @nox

**Comment by Manishearth** [comment]
r=me if you can figure out that crash

**Comment by KiChjang** [comment]
There's still one failing test in `/html/semantics/tabular-data/the-table-element/table-rows.html`, and it's because I did not preserve the ordering of the TR elements. The spec states that .rows should always return an HTMLCollection of TRs in thead first, followed by TRs in tbody or orphaned in table, followed by TRs in tfoot. I don't think HTMLCollection currently possess an API that allows me to specify the ordering of its elements.

**Comment by Manishearth** [comment]
Ah, right, in that case we should get that support in. Just allow arbitrary `Box<Item=Iterator>` in HTMLCollection with a utility method to use a filter to create an iterator.

**Comment by KiChjang** [comment]
Don't merge this yet, I've still yet to figure out why it's crashing and I still need to solve the one last test failure.

**Comment by KiChjang** [comment]
Actually, I should merge this and open up a follow-up issue to fix the `HTMLCollection` iterator stuff. Will look at the crashing issue.

File: components/script/dom/htmltableelement.rs [hunk 3]
```diff
@@ -51,6 +54,31 @@ impl HTMLTableElement {
 }
 
 impl HTMLTableElementMethods for HTMLTableElement {
+    // https://html.spec.whatwg.org/multipage/#dom-table-rows
+    fn Rows(&self) -> Root<HTMLCollection> {
+        #[allow(moved_no_move)]
```
Comment (KiChjang) [inline components/script/dom/htmltableelement.rs]: My suspicion of why it's crashing is this here. `RootedVec` isn't supposed to be moved, but the `TableRowFilter` is moved to a new `HTMLCollection` below. I had to add this lint in order to silence the compiler error.

Comment (jdm) [inline components/script/dom/htmltableelement.rs]: Plausible! Can it store a reference instead?

Comment (jdm) [inline components/script/dom/htmltableelement.rs]: Never mind, that doesn't look like it would work.

Comment (KiChjang) [inline components/script/dom/htmltableelement.rs]: I'm not sure if there are any workarounds with this. If not, we'll have to properly refactor and implement how we want iterators to work in HTMLCollections.

Comment (jdm) [inline components/script/dom/htmltableelement.rs]: We can store a regular `Vec<JS<Node>>` in the filter, since it's reachable by the GC.

Comment (KiChjang) [inline components/script/dom/htmltableelement.rs]: I get this error
```
/home/keith/Workspace/servo/components/script/dom/htmltableelement.rs:60:13: 60:36 error: Type must be rooted, use #[must_root] on the struct definition to propagate, #[deny(unrooted_must_root)] on by default
/home/keith/Workspace/servo/components/script/dom/htmltableelement.rs:60             sections: Vec<JS<Node>>
                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~
```
Not sure where I can root this vec...

**Comment by KiChjang** [comment]
Wait nvm, Manishearth explicitly said to r=him when I can figure out the crash.

@bors-servo r=Manishearth

**Comment by KiChjang** [comment]
@bors-servo retry
* #10392, #8931

**Pull request merged**